### PR TITLE
(PUP-7845) Implement leaf certificate revocation checking

### PR DESCRIPTION
**Pull request by adrienthebo**
Puppet has always performed chain revocation checking when revocation
checking was enabled, but has never been able to load more than a single
CRL. Because of this Puppet could only be used with a root CA if
revocation checking was desired, and if intermediate CAs were in used
revocation had to be entirely disabled.

This commit sidesteps the issue of loading multiple CRLs by allowing
certificate revocation checking to be performed on the leaf certificate
only. This allows users to revoke certificates with an intermediate CA
that Puppet is managing, which is a definite improvement over not
allowing any revocation checking at all.

For the sake of compatibility certificate_revocation still defaults to
full chain checking; users must opt in to leaf revocation checking.

### CODE REVIEW

**Comment by adrienthebo** [comment]
I need to perform some FR before this gets merged, hold off on merging till that's handled.

File: lib/puppet/ssl/host.rb [hunk 1]
```diff
@@ -382,6 +368,33 @@ def state
       return 'revoked'
     end
   end
+
+  private
```
Comment (MosesMendoza) [inline lib/puppet/ssl/host.rb]: I think this makes this officially a backwards-incompatible change. Intentional?

Comment (MosesMendoza) [inline lib/puppet/ssl/host.rb]: oh never mind ignore me - this is a new method

File: lib/puppet/ssl/host.rb [hunk 2]
```diff
@@ -382,6 +368,33 @@ def state
...
+      if crl = Puppet::SSL::CertificateRevocationList.indirection.find(CA_NAME)
+        flags = OpenSSL::X509::V_FLAG_CRL_CHECK
+        if Puppet.settings[:certificate_revocation] == :chain
+          flags |= OpenSSL::X509::V_FLAG_CRL_CHECK_ALL
+        end
+
+        store.flags = flags
+        store.add_crl(crl.content)
+      else
+        Puppet.debug _("Certificate revocation checking is enabled but a CRL cannot be found; CRL checking will not be performed.")
```
Comment (MosesMendoza) [inline lib/puppet/ssl/host.rb]: Oh nice addition

File: lib/puppet/defaults.rb [hunk 3]
```diff
@@ -854,11 +854,23 @@ def self.default_diffargs
...
-        :desc     => "Whether certificate revocation should be supported by downloading a
-          Certificate Revocation List (CRL)
-          to all clients.  If enabled, CA chaining will almost definitely not work.",
+        :default  => 'chain',
+        :type     => :certificate_revocation,
+        :desc     => <<EOT
+Whether certificate revocation checking should be enabled, and what level of checking should be performed.
+
+When certificate_revocation is set to 'true' or 'chain', Puppet will download the CA CRL and will perform revocation
+checking against each certificate in the chain. When Puppet is using a single root certificate and has no intermediates
```
Comment (MosesMendoza) [inline lib/puppet/defaults.rb]: > Puppet will download the CA CRL and will perform revocation checking against each certificate in the chain. 

To rephrase this to confirm I understand what we mean, would this be the same as saying:
"Puppet will download the CA CRL and will check each certificate in the chain against that CRL" ?

File: lib/puppet/defaults.rb [hunk 4]
```diff
@@ -854,11 +854,23 @@ def self.default_diffargs
         This is distinct from the certificate authority's CRL."
     },
     :certificate_revocation => {
-        :default  => true,
-        :type     => :boolean,
-        :desc     => "Whether certificate revocation should be supported by downloading a
-          Certificate Revocation List (CRL)
-          to all clients.  If enabled, CA chaining will almost definitely not work.",
+        :default  => 'chain',
```
Comment (MosesMendoza) [inline lib/puppet/defaults.rb]: I'm curious why `:chain` ? I'm reading this PR as `true` and `:chain` are synonyms, and to fulfill the purpose of the PR all we strictly need is a new value `:leaf`. What's the direction behind `:chain` ?

File: lib/puppet/defaults.rb (hunk 3 above)
Comment (adrienthebo) [inline lib/puppet/defaults.rb]: Mostly; there should be a CRL for each CA certificate in the chain and Puppet will consult the corresponding CRL for each CA certificate. It's a little bit tricky to phrase.

File: lib/puppet/defaults.rb (hunk 4 above)
Comment (adrienthebo) [inline lib/puppet/defaults.rb]: This matches the semantics of the Apache [`SSLCARevocationCheck`](https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslcarevocationcheck) setting that has `chain`, `leaf`, and `none` values. Since this value is no longer a boolean value I would rather be clear about what sort of validation we're doing, IE chain vs leaf validation, rather than have a weird ternary `true`|`leaf`|`false` triple.

Comment (adrienthebo) [inline lib/puppet/defaults.rb]: See also: https://thedailywtf.com/articles/What_Is_Truth_0x3f_

File: lib/puppet/defaults.rb (hunk 3 above)
Comment (MosesMendoza) [inline lib/puppet/defaults.rb]: Ah and this is in the world where the user must have manually distributed the individual CRLs because we can't?

Comment (adrienthebo) [inline lib/puppet/defaults.rb]: This is that world; unfortunately that world is still broken because we can't load multiple CRLs. My intention with this was to maintain the current behavior to reduce surprises, and when we can support multiple CRLs the chain cert verification will start working. This is a small amount of paranoia and a larger amount of future proofing.

Comment (MosesMendoza) [inline lib/puppet/defaults.rb]: Ah ok. So if I understand correctly this sentence:
> Puppet will download the CA CRL and will perform revocation checking against each certificate in the chain.

is not entirely currently true, because CRL checking against multiple CRLs doesn't currently work? 

If that's true, could we reword this to that effect? Something like:
```
When certificate revocation is set to 'true' or 'chain', Puppet will download the CA CRL and perform revocation checking against the CA cert. However, the CA cert cannot contain intermediate certificates when these settings are in use. If intermediate CAs are in use, revocation checking will fail due to current limitations in loading multiple CRLs. Thus the value of 'chain' is optimistic - it is a goal to support revocation checking of chained CAs with multiple associated CRLs in the future.
```

Comment (adrienthebo) [inline lib/puppet/defaults.rb]: Whether we should update this comment depends on if/when we're going to fix chain CRL verification, and whether we need to expose the details of CRL checking to most users who will be using a root CA certificate. My approach was that this behavior has been an issue since CRL checking was implemented so we should generally preserve the existing code paths, while optimistically planning for actually fixing this. if it's likely that we're not going to implement chain verification then we should add the caveats that you've written.

Comment (MosesMendoza) [inline lib/puppet/defaults.rb]: Agreed this has long been both incomplete and inaccurate. That said I advocate for having the docs/info as accurate as possible for the code they represent at any given time so since we're in here making improvements we might as well be honest about the state of affairs. We can make further corrections as they're made in the code?

Comment (adrienthebo) [inline lib/puppet/defaults.rb]: Will do.

**Comment by adrienthebo** [comment]
I've performed FR on this with mixed results; leaf CRL checking does work if the CRL is present. However my work uncovered another issue in that Puppet can recurse infinitely when trying to fetch the CRL, no CRL is present, and CRL verification is enabled. If the CRL is manually copied in place then revocation checking works This PR can be merged but I'll need to do some follow up work to resolve the recursion issue. I can do that today.

**Comment by MosesMendoza** [comment]
Adding blocked label until we circle back with @adrienthebo on next steps

**Comment by adrienthebo** [comment]
This ticket should be unblocked. /cc @MosesMendoza @justinstoller

File: lib/puppet/network/http/connection.rb [hunk 5]
```diff
@@ -222,15 +222,15 @@ def with_connection(site, &block)
...
-        raise Puppet::Error, msg, error.backtrace
-      elsif error.message.include? "certificate verify failed"
+      if error.message.include? "certificate verify failed"
         msg = error.message
         msg << ": [" + @verify.verify_errors.join('; ') + "]"
         raise Puppet::Error, msg, error.backtrace
+      elsif peer_cert && !OpenSSL::SSL.verify_certificate_identity(peer_cert.content, site.host)
+        valid_certnames = [peer_cert.name, *peer_cert.subject_alt_names].uniq
+        certnames_msg = valid_certnames.length > 1 ? _("one of %{certnames}") % { certnames: valid_certnames.join(', ') } : valid_certnames.first
+        msg = _("Server hostname '%{host}' did not match server certificate; expected %{msg}") % { host: site.host, msg: certnames_msg }
```
Comment (Magisus) [inline lib/puppet/network/http/connection.rb]: This kind of message that we piece together this way is very hard to translate. I think the clearest way to clean this up would be to include the "expected" in the smaller strings above, someting like:
```valid_certnames.length > 1 ? _("expected one of %{certnames}") % { certnames: valid_certnames.join(', ') } : _("expected %{certname}") % { certname: valid_certnames.first }
```

Comment (adrienthebo) [reply lib/puppet/network/http/connection.rb]: Done.

File: lib/puppet/defaults.rb (hunk 3 above)
Comment (adrienthebo) [reply lib/puppet/defaults.rb]: Documentation updated to reflect reality.
